File: civitasJAVA/civitas/src/civitas/Dado.java
package civitas;
import java.util.Random;
import java.util.ArrayList;


public class Dado {
  
    private Random random = new Random();
    private int ultimoResultado;
    private Boolean debug;
    private static final Dado instance = new Dado();
    private static int SalidaCarcel = 5;
    

    //Constructor privado
    private Dado(){
        debug = false;
        ultimoResultado = 0;
        random = new Random();
    }
    
    //Implementación de clase Singleton
    static Dado getInstance() { return instance; }
    
    //Devuelve número aleatorio de tirada
    int tirar(){
        int resultado;
        if(debug == false)
            resultado = random.nextInt(6)+1;
        else 
            resultado = 1;

        ultimoResultado = resultado;
        return resultado;
    }
   
    //Determina aleatoriamente si se sale de la cárcel
    boolean salgoDeLaCarcel(){
        Boolean resultado = false;
        if(getUltimoResultado()==6)
            resultado = true;

        return resultado;
    }
       
    //Dertermina aleatoriamente quién empieza
    int quienEmpieza(int n){
        return  random.nextInt(n);
    }
    
    //Devuelve el último resultado que arrojó el dado
    int getUltimoResultado(){
        return ultimoResultado;
    }
   
    //Setter de modo debug
    public void setDebug(Boolean d) {
        debug = d;
        if(debug==true)
            Diario.getInstance().ocurreEvento("Dado debug true");
        else 
            Diario.getInstance().ocurreEvento("Dado debug false");
    }
   
}

    

File: civitasRUBY/lib/dado.rb
#encoding: utf-8

require 'singleton'
require 'diario'

module Civitas
  class Dado
    include Singleton #Solo existe una instancia de Dado
    
    attr_reader :ultimoResultado
    
    @@SalidaCarcel = 5
    
    #Initialize
    def initialize
      @random          = Random.new()
      @ultimoResultado = 0      #Inicializada por iniciativa propia
      @debug           = false  #Inicializada por iniciativa propia
    end
    
    #Devuelve tirada [1,6] o 1 en debug
    def tirar
      if !@debug
        @ultimoResultado = @random.rand(6)+1
      else
        @ultimoResultado = 1
      end
      
      return @ultimoResultado
    end
    
    #Determina si sale de la carcel con una tirada
    def salgoDeLaCarcel
     
      if @ultimoResultado ==6
        return true
      else
        return false
      end
    end
    
    #Generador aleatorio que determina cual de los n jugadores empieza
    def quienEmpieza(n)
      return @random.rand(n)
    end
    
    #Devueleve el último valor arrojado por el dado
    def ultimoResultado
      return ultimoResultado
    end
    
    #Set de @debug
    def debug(_debug)
      @debug = _debug
     
      if @debug == true
        Diario.instance.ocurreEvento("Activacion modo debug")
      else 
        Diario.instance.ocurreEvento("Desactivacion modo debug")
      end
      
    
    end
    
  end
      private :initialize
  
end
